File: c_files/text_games/dungeon_delve/src/enemies.c
#include "enemies.h"

void enemy_init(Enemy *enemy, int level, EnemyType type) {
  enemy->type = type;
  enemy->level = level;

  if (enemy->type == ZOMBIE) {
    enemy->base_health = 50;
    enemy->base_damage = 5;
    enemy->health_growth_rate = 1.10f;
    enemy->damage_growth_rate = 1.05f;
  }
  if (enemy->type == GOBLIN) {
    enemy->base_health = 40;
    enemy->base_damage = 7;
    enemy->health_growth_rate = 1.08f;
    enemy->damage_growth_rate = 1.06f;
  }
  if (enemy->type == SKELETON) {
    enemy->base_health = 45;
    enemy->base_damage = 6;
    enemy->health_growth_rate = 1.09f;
    enemy->damage_growth_rate = 1.05f;
  }
  if (enemy->type == ORC) {
    enemy->base_health = 70;
    enemy->base_damage = 10;
    enemy->health_growth_rate = 1.12f;
    enemy->damage_growth_rate = 1.08f;
  }
  if (enemy->type == TROLL) {
    enemy->base_health = 100;
    enemy->base_damage = 8;
    enemy->health_growth_rate = 1.15f;
    enemy->damage_growth_rate = 1.04f;
  }
  if (enemy->type == VAMPIRE) {
    enemy->base_health = 65;
    enemy->base_damage = 12;
    enemy->health_growth_rate = 1.10f;
    enemy->damage_growth_rate = 1.10f;
  }
  if (enemy->type == WEREWOLF) {
    enemy->base_health = 80;
    enemy->base_damage = 11;
    enemy->health_growth_rate = 1.12f;
    enemy->damage_growth_rate = 1.09f;
  }
  if (enemy->type == DEMON) {
    enemy->base_health = 90;
    enemy->base_damage = 14;
    enemy->health_growth_rate = 1.13f;
    enemy->damage_growth_rate = 1.12f;
  }
  if (enemy->type == DRAGON) {
    enemy->base_health = 150;
    enemy->base_damage = 20;
    enemy->health_growth_rate = 1.18f;
    enemy->damage_growth_rate = 1.15f;
  }
  if (enemy->type == BANDIT) {
    enemy->base_health = 55;
    enemy->base_damage = 9;
    enemy->health_growth_rate = 1.09f;
    enemy->damage_growth_rate = 1.07f;
  }
  if (enemy->type == WITCH) {
    enemy->base_health = 60;
    enemy->base_damage = 13;
    enemy->health_growth_rate = 1.08f;
    enemy->damage_growth_rate = 1.13f;
  }
}

bool enemy_is_alive(Enemy *enemy) { return enemy->base_health > 0; }

File: python_files/text_games/dungeon_delve/enemies.py
from abc import ABC


class Enemy(ABC):
    def __init__(self, x: int, y: int, level: int) -> None:
        self.x: int = x
        self.y: int = y
        self.level: int = level
    
    def is_alive(self) -> bool:
        return self.base_health > 0


class Zombie(Enemy):
    def __init__(self, x, y, level) -> None:
        super().__init__(x, y, level)

        self.base_health: int = 50
        self.base_damage: int = 5
        self.health_growth_rate: int = 1.10
        self.damage_growth_rate: int = 1.05


class Goblin(Enemy):
    def __init__(self, x, y, level) -> None:
        super().__init__(x, y, level)
        self.base_health: int = 40
        self.base_damage: int = 7
        self.health_growth_rate: int = 1.08
        self.damage_growth_rate: int = 1.06


class Skeleton(Enemy):
    def __init__(self, x, y, level) -> None:
        super().__init__(x, y, level)
        self.base_health: int = 45
        self.base_damage: int = 6
        self.health_growth_rate: int = 1.09
        self.damage_growth_rate: int = 1.05


class Orc(Enemy):
    def __init__(self, x, y, level) -> None:
        super().__init__(x, y, level)
        self.base_health: int = 70
        self.base_damage: int = 10
        self.health_growth_rate: int = 1.12
        self.damage_growth_rate: int = 1.08


class Troll(Enemy):
    def __init__(self, x, y, level) -> None:
        super().__init__(x, y, level)
        self.base_health: int = 100
        self.base_damage: int = 8
        self.health_growth_rate: int = 1.15
        self.damage_growth_rate: int = 1.04


class Vampire(Enemy):
    def __init__(self, x, y, level) -> None:
        super().__init__(x, y, level)
        self.base_health: int = 65
        self.base_damage: int = 12
        self.health_growth_rate: int = 1.10
        self.damage_growth_rate: int = 1.10


class Werewolf(Enemy):
    def __init__(self, x, y, level) -> None:
        super().__init__(x, y, level)
        self.base_health: int = 80
        self.base_damage: int = 11
        self.health_growth_rate: int = 1.12
        self.damage_growth_rate: int = 1.09


class Demon(Enemy):
    def __init__(self, x, y, level) -> None:
        super().__init__(x, y, level)
        self.base_health: int = 90
        self.base_damage: int = 14
        self.health_growth_rate: int = 1.13
        self.damage_growth_rate: int = 1.12


class Dragon(Enemy):
    def __init__(self, x, y, level) -> None:
        super().__init__(x, y, level)
        self.base_health: int = 150
        self.base_damage: int = 20
        self.health_growth_rate: int = 1.18
        self.damage_growth_rate: int = 1.15


class Bandit(Enemy):
    def __init__(self, x, y, level) -> None:
        super().__init__(x, y, level)
        self.base_health: int = 55
        self.base_damage: int = 9
        self.health_growth_rate: int = 1.09
        self.damage_growth_rate: int = 1.07


class Witch(Enemy):
    def __init__(self, x, y, level) -> None:
        super().__init__(x, y, level)
        self.base_health: int = 60
        self.base_damage: int = 13
        self.health_growth_rate: int = 1.08
        self.damage_growth_rate: int = 1.13
